firewalldb/sql_migration_test.go

```diff
@@ -0,0 +1,666 @@
+	)
+
+	fEntries4 := insertTempAndPermEntry(
+		t, ctx, boltDB, testRuleName2, groupAlias,
+		fn.Some(testFeatureName), testEntryKey2, nilValue,
+	)
+	fEntries5 := insertTempAndPermEntry(
+		t, ctx, boltDB, testRuleName2, groupAlias,
+		fn.Some(testFeatureName), testEntryKey3, emptyValue,
+	)
+
+	return mergeEntries(
+		gEntries1, gEntries2, gEntries3, gEntries4, sEntries1,
+		sEntries2, sEntries3, sEntries4, fEntries1, fEntries2,
+		fEntries3, fEntries4, fEntries5,
+	)
+}
+
+func getNewSessionAlias(t *testing.T, ctx context.Context,
+	sessionStore session.Store) fn.Option[[]byte] {
+
+	sess, err := sessionStore.NewSession(
+		ctx, "test", session.TypeAutopilot,
+		time.Unix(1000, 0), "something",
+	)
+	require.NoError(t, err)
+
+	return fn.Some(sess.GroupID[:])
+}
+
+// insertTempAndPermEntry populates the kv store with one entry for the local
+// temp store, and one entry for the local store. Both of the entries will be
+// inserted with the same groupAlias, ruleName, entryKey and entryValue.
+func insertTempAndPermEntry(t *testing.T, ctx context.Context,
+	boltDB *BoltDB, ruleName string, groupAlias fn.Option[[]byte],
+	featureNameOpt fn.Option[string], entryKey string,
+	entryValue []byte) []*kvEntry {
+
+	tempKvEntry := &kvEntry{
+		ruleName:    ruleName,
+		groupAlias:  groupAlias,
+		featureName: featureNameOpt,
+		key:         entryKey,
+		value:       entryValue,
+		perm:        false,
+	}
+
+	insertKvEntry(t, ctx, boltDB, tempKvEntry)
+
+	permKvEntry := &kvEntry{
+		ruleName:    ruleName,
+		groupAlias:  groupAlias,
+		featureName: featureNameOpt,
+		key:         entryKey,
+		value:       entryValue,
+		perm:        true,
+	}
+
+	insertKvEntry(t, ctx, boltDB, permKvEntry)
+
+	return []*kvEntry{tempKvEntry, permKvEntry}
+}
+
+// insertKvEntry populates the kv store with passed entry, and asserts that the
+// entry is inserted correctly.
+func insertKvEntry(t *testing.T, ctx context.Context,
+	boltDB *BoltDB, entry *kvEntry) {
+
+	if entry.groupAlias.IsNone() && entry.featureName.IsSome() {
+		t.Fatalf("cannot set both global and feature specific at the " +
+			"same time")
+	}
+
+	// We get the kv stores that the entry will be inserted into. Note that
+	// we set an empty group ID if the entry is global, as the group ID
+	// will not be used when fetching the actual kv store that's used for
+	// global entries.
+	groupID := [4]byte{}
+	if entry.groupAlias.IsSome() {
+		copy(groupID[:], entry.groupAlias.UnwrapOrFail(t))
+	}
+
+	kvStores := boltDB.GetKVStores(
+		entry.ruleName, groupID, entry.featureName.UnwrapOr(""),
+	)
+
+	err := kvStores.Update(ctx, func(ctx context.Context,
+		tx KVStoreTx) error {
+
+		var store KVStore
+		switch {
+		case entry.groupAlias.IsNone() && entry.perm:
+			store = tx.Global()
+		case entry.groupAlias.IsNone() && !entry.perm:
+			store = tx.GlobalTemp()
+		case entry.groupAlias.IsSome() && !entry.perm:
+			store = tx.LocalTemp()
+		case entry.groupAlias.IsSome() && entry.perm:
+			store = tx.Local()
+		}
```

Comment:
note that in the example i posted before, we could not possibly end up with a nil store. So if you want to have these 4 cases and not pre assigne a default, the rather have a default case to catch that none of the cases match